Add DAP ENV transformation code

quazi-h opened the pull request:
## Why

[Relevant ticket](https://broadinstitute.atlassian.net/browse/1287)
[Bug ticket](https://broadinstitute.atlassian.net/browse/1370)

This is the third and final PR in a series to get the DAP environmental data extracted from RedCap.
We are breaking a larger more convoluted PR into smaller pieces and this is the last of that series.
We are also addressing a duplicate records issue in this PR.

## This PR
-Added 5 transformation scripts for the env data subsets
-Added a unit test for each type of env transformation
-Refactored HLESurveyTransformationPipelineBuilder to process environmental records separately
-Updated the hardcoded list of arms to include largest subset we can grab successfully
-Addressed duplicates (DSPDC-1370) in data by adding some logic to RawRecord.
-Added logic to process the Environment table in the tsv script under /hack
-Expanded the list of arms to environment to include the full set of events in RedCap we should be pulling.
-Modified the client code to take "arm" as a param
-Added a filter for "NA" values when getOptionalDateTime is used
-Added logic to remove "NA" values from walkscoreDates
-Added unit test cases for new logic and fixed some unit test bugs
-Adjusted the idBatchSize for Environment to 10 to accommodate RedCap
-Split up the pull for initRequests to pull a list of records per arm

quazi-h committed:
git commit -m "Extend the transformation pipeline code to process the environment data
Add environment transformation code
Modified RawRecord to include a few additional get methods
Dedupe rows by converting the values pulled for each field to a set and grabbing the head"

quazi-h committed:
Merge branch 'master' into QH-DSPDC-1287-add-env-transformations

quazi-h committed:
Adding individual transformation scripts for env subsets

quazi-h committed:
add unit tests

quazi-h committed:
Merge branch 'master' into QH-DSPDC-1287-add-env-transformations

quazi-h committed:
"Updated hardcoded list of arms"

quazi-h committed:
Reverting some changes to CSLB that were not intended for this PR

aherbst-broad commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/HLESurveyTransformationPipelineBuilder.scala:
```diff
@@ -77,4 +86,34 @@ object HLESurveyTransformationPipelineBuilder extends PipelineBuilder[Args] {
+      .groupBy(foo => {
+        (foo.read[String]("record"), foo.read[String]("redcap_event_name"))
+      })
+      //.groupBy(read[String]("redcap_event_name"))
```
Remove this commented out code

aherbst-broad commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/HLESurveyTransformationPipelineBuilder.scala:
```diff
@@ -77,4 +86,34 @@ object HLESurveyTransformationPipelineBuilder extends PipelineBuilder[Args] {
+    // to get the format: (studyId, arm_id, Iterable((fieldName, Iterable(value))))
+    // (study_id, arm_id, iterable())
+    rawRecords
+      .groupBy(foo => {
```
🔧 Can you give this a more descriptive name than `foo`? Perhaps `record`?

aherbst-broad commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/HLESurveyTransformationPipelineBuilder.scala:
```diff
@@ -77,4 +86,34 @@ object HLESurveyTransformationPipelineBuilder extends PipelineBuilder[Args] {
+              case (fieldName, rawValues) =>
+                (fieldName, rawValues.map(_.read[String]("value")).toArray.sorted)
+            }
+          val bar = fields + ("redcap_event_name" -> Array(eventName))
```
🔧 Same comment re: naming; you might do away with this intermediate val altogether and do the concat inline when you construct the `RawRecord` instance.

aherbst-broad commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/RawRecord.scala:
```diff
@@ -22,7 +22,7 @@ case class RawRecord(id: Long, fields: Map[String, Array[String]]) {
   def getRequired(field: String): String = {
     val values = fields.getOrElse(field, Array.empty)
-    if (values.length != 1) {
+    if (values.toSet.size != 1) {
```
❤️

aherbst-broad replied:
Can you add a test somewhere that exercises this de-duping behavior?

aherbst-broad commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/environment/WalkabilityTransformations.scala:
```diff
@@ -0,0 +1,34 @@
+    if (!rawRecord.getOptional("wv_walkscore").contains("NA")) {
+      walkscore = rawRecord.getOptional("wv_walkscore").map(_.toDouble)
+    }
+    if (rawRecord.getArray("wv_walkscore_date").nonEmpty) {
```
Can you add a comment describing why this nested conditional exists (I think I remember from our pairing, but we will for sure forget a month from now).

quazi-broad replied:
We observed instances of "NA" in data returned from RedCap - thanks I will add the comment so we don't scratch our heads in the future

ThePletch replied:
could we include handling for `"NA"` values in our various `getOptional/getArray/getWhatever` methods rather than explicitly handling that here, or would that potentially swallow cases where `"NA"` is a valid value for a field?

aherbst-broad commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/environment/WalkabilityTransformations.scala:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    EnvironmentWalkability(
+      wvWalkscore = walkscore, // rawRecord.getOptional("wv_walkscore").map(_.toDouble),
```
🔧 Remove commented out code.

aherbst-broad commented on hles/transformation/src/test/scala/org/broadinstitute/monster/dap/environment/WalkabilityTransformationsSpec.scala:
```diff
@@ -0,0 +1,37 @@
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+class WalkabilityTransformationsSpec extends AnyFlatSpec with Matchers with OptionValues {
```
🔧 I think this test should also exercise the conditional codepaths around `wv_walkscore_date` and `wv_walkscore`

aherbst-broad commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/HLESurveyTransformationPipelineBuilder.scala:
```diff
@@ -5,6 +5,7 @@ import com.spotify.scio.values.SCollection
 import org.slf4j.{Logger, LoggerFactory}
+import upack.Msg
 
 object HLESurveyTransformationPipelineBuilder extends PipelineBuilder[Args] {
```
🌱 Out of scope for this PR but: we need to do some refactoring after this ships, we're now performing HLES, CSLB and env transforms in a class that is ostensibly for HLES transforms (from it's name at least)

quazi-broad replied:
agreed - I really wanted to break out the transformation into 3 streams in this PR but figured we should keep it simple since this is already 1/3 of a larger PR and there's a bunch of stuff being added in this PR alone

aherbst-broad commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/HLESurveyTransformationPipelineBuilder.scala:
```diff
@@ -77,4 +86,34 @@ object HLESurveyTransformationPipelineBuilder extends PipelineBuilder[Args] {
+    val rawRecords: SCollection[Msg] = StorageIO
+      .readJsonLists(
+        ctx,
+        "Raw Records",
```
⛏️ "Raw Environmental Records" perhaps?

aherbst-broad reviewed (commented):
Looking good so far!

aherbst-broad commented on hles/extraction/src/main/scala/org/broadinstitute/monster/dap/EnvironmentExtractionPipeline.scala:
```diff
@@ -24,7 +24,23 @@ object EnvironmentExtractionPipeline extends ScioApp[Args] {
+      "may2020_arm_1",
+      "may2020_secondary_arm_1",
+      "june2020_arm_1",
+      "june2020_secondary_arm_1"
```
❓ Does this only go to June 2020 for performance reasons?

quazi-broad replied:
yes - adding any more arms makes this fail when running a full refresj

ThePletch replied:
do we have any mechanisms to throttle this to let us add all of the arms we need without beating redcap into submission? this seems like a high priority feature to add if it's stopping us from doing a proper full refresh.

aherbst-broad replied:
No, there is no such mechanism now; @quazi-broad and I have been talking about this, we'll tackle it after this gets merged (@quazi-broad can you make sure there's a ticket for this and link it here?)

quazi-broad replied:
https://broadinstitute.atlassian.net/browse/DSPDC-1415

quazi-h committed:
Addressing PR comments
More descriptive variable names, code refactors, adding a unit test case for some new logic in walkability data

quazi-h committed:
Adding some comments on the deduplication logic and added a unit test that check that logic.

ThePletch commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/EnvironmentTransformations.scala:
```diff
@@ -0,0 +1,27 @@
+    * The schema for environment variables has been separated into 5 jade-fragments.
+    */
+  def mapEnvironment(rawRecord: RawRecord): Option[Environment] = {
+    val dog_id = rawRecord.id
```
:snake: -> :camel:
```suggestion
    val dogId = rawRecord.id
```

ThePletch commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/EnvironmentTransformations.scala:
```diff
@@ -0,0 +1,27 @@
+      Environment(
+        dogId = dog_id,
+        addressMonthYear = rawRecord.getRequired("redcap_event_name"),
+        environmentGeocoding = Some(GeocodingTransformations.mapGeocodingMetadata(rawRecord)),
```
for my understanding, what's the purpose of wrapping these in `Some`?

quazi-broad replied:
To get the Option[] type, I believe

ThePletch commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/HLESurveyTransformationPipelineBuilder.scala:
```diff
@@ -77,4 +86,32 @@ object HLESurveyTransformationPipelineBuilder extends PipelineBuilder[Args] {
+            .groupBy(_.read[String]("field_name"))
+            .map {
+              case (fieldName, rawValues) =>
+                (fieldName, rawValues.map(_.read[String]("value")).toArray.sorted)
```
this block (converting from redcap data structures to something more useful to us) seems like it's something we might want to be able to reuse. can we pull this functionality out into a `RedcapRecordDigester` (working title) object or something similar?

ThePletch commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/RawRecord.scala:
```diff
@@ -33,10 +37,13 @@ case class RawRecord(id: Long, fields: Map[String, Array[String]]) {
     * Raises an error if the attribute has multiple values.
+    *
+    * Only unique duplicates will throw an error
+    * If we receive multiple fields with the same value, the head of the set is used
```
is it necessarily safe to discard the extra data if we get multiple unique values for a field?

quazi-broad replied:
I think the way we have this written - we will throw an error from unique duplicates.
This code should only weed out instances with the same value.

ThePletch commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/environment/CensusTransformations.scala:
```diff
@@ -0,0 +1,43 @@
+      cvPctOwnerOccupied = rawRecord.getOptional("cv_pctowneroccupied").map(_.toDouble),
+      cvPctUsBorn = rawRecord.getOptional("cv_pctborninus").map(_.toDouble),
+      cvStabilityIndex = rawRecord.getOptional("cv_stabilityindex").map(_.toDouble),
+      cvDataYear = rawRecord.getOptionalNumber("cv_data_year")
```
definitely not in scope for this PR, but this style of transformation definition feels _really_ verbose. finding myself wondering if we could define transformations more declaratively than this, e.g. just `Map("cv_pctowneroccupied" -> Option[Double], ...)`

ThePletch commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/environment/TemperaturePrecipitationTransformations.scala:
```diff
@@ -0,0 +1,116 @@
+      tpPcpnAnnual09 = rawRecord.getOptional("tp_pcpn_annual_09").map(_.toDouble),
+      tpPcpnAnnual10 = rawRecord.getOptional("tp_pcpn_annual_10").map(_.toDouble),
+      tpPcpnAnnual11 = rawRecord.getOptional("tp_pcpn_annual_11").map(_.toDouble),
+      tpPcpnAnnual12 = rawRecord.getOptional("tp_pcpn_annual_12").map(_.toDouble),
```
is it important that we define these individually in sequence, or is there some way we could just pass in all of these values in a loop for each month, 1-12?

aherbst-broad replied:
This could get sticky; we codegen (via a plugin from `ingest-utils`) the underlying case class `EnvironmentTemperaturePrecipitation` from a JSON schema. We could muck with the case class JSON encoder to do the array transformation as you allude to, but it'd get clobbered on the next compile.

ThePletch commented on hles/transformation/src/test/scala/org/broadinstitute/monster/dap/environment/TemperaturePrecipitationTransformationsSpec.scala:
```diff
@@ -0,0 +1,223 @@
+      "tp_tmin_norm_09" -> Array("47.1"),
+      "tp_tmin_norm_10" -> Array("40.5"),
+      "tp_tmin_norm_11" -> Array("34.9"),
+      "tp_tmin_norm_12" -> Array("30.3")
```
at least in the specs, i think we should definitely be building these massive maps programmatically

ThePletch commented on hles/transformation/src/test/scala/org/broadinstitute/monster/dap/environment/WalkabilityTransformationsSpec.scala:
```diff
@@ -0,0 +1,56 @@
+    // output of the example record's walkability transformations
+    walkabilityDataMapped.wvWalkscore shouldBe Some(76.0)
+    walkabilityDataMapped.wvWalkscoreDescrip.value shouldBe 2L
+    // todo: CHECK THIS - CONFIRM WITH DAP ON TYPE
```
flagging this todo to be addressed before merging

ThePletch reviewed (commented):
couple of minor nits, questions, and down-the-line refactor suggestions

quazi-h committed:
Style fix - switch to camel case.
Updated wv_walkscore_date to be a date instead of a string.
Fixed DateTimeFormatter string (DAP uses 24 hour vs 12 hour).
Updated WalkabilityTransformations to call that custom date formatter that was added to RawRecord.

quazi-h committed:
Added additional test cases for walkability transformations - this exposed a potential bug (if there's multiple values for a field and one of the values is "NA", it will throw an IllegalStateException.
Here is my workaround (please see how I modified getOptional in RawRecord.scala. It is filtering out "NA" when there is more than one value - the only instance in which this could harm us is maybe if there was multiple valid values and one of them is actually "NA"?`

aherbst-broad committed:
Add http logging interceptor to help with redcap debugging, as well as capturing the response body on parsing error

aherbst-broad committed:
Merge branch 'arh-add-http-logging-interceptor' into arh-env-hacking

aherbst-broad committed:
Dispatch arms individually

aherbst-broad committed:
Dispatch arms individually (#91)

aherbst-broad committed:
Add handling for multiple datetimes that differ only in microseconds, with test.

aherbst-broad committed:
Merge branch 'QH-DSPDC-1287-add-env-transformations' of github.com:DataBiosphere/dog-aging-ingest into QH-DSPDC-1287-add-env-transformations

quazi-h committed:
merge branch 'arh-env-hacking' into QH-DSPDC-1287-add-env-transformations
added okHttp logging interceptor

quazi-h committed:
Merge branch 'master' into QH-DSPDC-1287-add-env-transformations

quazi-h committed:
Added logic to process the Environment table in the tsv script under /hack
Expanded the list of arms to environment to include the full set of events in RedCap we should be pulling.
Had to modify the client code to take "arm" as a param
Added a filter for "NA" values when getOptionalDateTime is used
Added logic to remove "NA" values from walkscoreDates
Added unit test cases for new logic and fixed some unit test bugs

quazi-h committed:
Adjusting the idBatchSize for Environment to 10 to accommodate RedCap
Renamed "arms" variable, splitting up the pull for initRequests to pull a list of records per arm

quazi-h committed:
Merge branch 'master' into QH-DSPDC-1287-add-env-transformations

aherbst-broad committed:
Fix extractionpipelinebuilderspec

aherbst-broad reviewed (approved):
lgtm

ThePletch commented on hack/convert-output-to-tsv.py:
```diff
@@ -31,134 +31,149 @@ def printd(x):
+                    dw.writeheader()
+                    dw.writerows(row_list)
+                print(f"...{table_name} was successfully written to {output_location}")
+    except OSError as e:
```
:-1: big try-catch blocks like this are a code smell and can make issues harder to debug. let's try to identify lines where we expect errors and want to skip over them (do we, since all we're doing is printing the error?) and put try-catch blocks around them instead.

quazi-broad replied:
I've walked back these changes minus just adding the environment table logic.

ThePletch commented on hles/extraction/src/main/scala/org/broadinstitute/monster/dap/EnvironmentExtractionPipeline.scala:
```diff
@@ -35,7 +62,7 @@ object EnvironmentExtractionPipeline extends ScioApp[Args] {
       fieldList,
       subdir,
-      100,
+      10,
```
let's add a comment about what led us to pick a value of 10 here.

ThePletch commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/RawRecord.scala:
```diff
@@ -36,11 +42,20 @@ case class RawRecord(id: Long, fields: Map[String, Array[String]]) {
+      if (values.--(Array("NA")).size > 1) {
+        throw new IllegalStateException(s"Record $id has multiple values for field $field")
+      } else {
+        values.headOption
```
are we worried about this possibly returning `NA`?

quazi-broad replied:
Yes, we were seeing `NA` values being pulled by getOption

ThePletch commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/RawRecord.scala:
```diff
@@ -52,11 +67,13 @@ case class RawRecord(id: Long, fields: Map[String, Array[String]]) {
     getOptional(field).map(value => {
       try {
-        value.toLong
+        value.toDouble.toLong
```
:-1: if we convert to a double in this try statement, it'll also handle any decimal values, which will prevent us from logging that we needed to truncate a decimal value and instead simply do so silently.

quazi-broad replied:
Moved the toDouble into the TruncatedDecimalError logic

ThePletch reviewed (changes requested):
one thing that we need to change to avoid clobbering existing functionality, then a few suggestions for improvements

quazi-h committed:
TSV script will be getting an overhaul from Steve, walking back try, catch statement
Moved the toDouble conversion into the truncatedValue error handling block (since my change would allow floats to get into the data).
Added some descriptive code comments.

quazi-h committed:
Merge branch 'QH-DSPDC-1287-add-env-transformations' of github.com:DataBiosphere/dog-aging-ingest into QH-DSPDC-1287-add-env-transformations

quazi-h committed:
walking back tsv script changes
